client/src/main/java/org/eclipse/hono/client/impl/HonoClientImpl.java

```diff
@@ -108,31 +108,25 @@ public void setCacheManager(final CacheManager manager) {
         this.cacheManager = Objects.requireNonNull(manager);
     }
 
-    /**
-     * Sets the connection to the Hono server.
-     * <p>
-     * This method is mostly useful to inject a (mock) connection when running tests.
-     *
-     * @param connection The connection to use.
-     */
-    void setConnection(final ProtonConnection connection) {
-        this.connection = connection;
+    @Override
+    public boolean isConnected() {
+        synchronized (connectionLock) {
+            return connection != null && !connection.isDisconnected();
+        }
     }
 
     /**
-     * Sets the vertx context to run all interactions with the Hono server on.
-     * <p>
-     * This method is mostly useful to inject a (mock) context when running tests.
+     * Sets the connection used to interact with the Hono server.
      *
-     * @param context The context to use.
+     * @param connection The connection to use.
+     * @param context The context that the connection has been established on.
+     *                It is important to use the connection on this context only.
      */
-    void setContext(final Context context) {
-        this.context = context;
-    }
-
-    @Override
-    public boolean isConnected() {
-        return connection != null && !connection.isDisconnected();
+    void setConnection(final ProtonConnection connection, final Context context) {
+        synchronized (connectionLock) {
```

Comment:
Make the whole method `synchronized`?
If using `this` as a lock, this should be the same semantic, shouldn't it?